Address PR #9175 review comments

emlun opened the pull request:
This is a meta-PR into:
- https://github.com/syncthing/syncthing/pull/9175

@imsodin I believe this should address these comments:

- https://github.com/syncthing/syncthing/pull/9175#discussion_r1700733287
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1700853007
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1700899146
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1712634228
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1712635690
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1712635805
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1700729175

emlun committed:
Remove leftover WebAuthn state management from adjustConfig

This should have been removed in 4a14a21a800242b05fb09262347868d3dc86efbf when
the WebAuthn state moved from config to DB.

emlun committed:
Extract type webauthnLibUser to implement webauthnLib.User

emlun committed:
Take GUIConfiguration parameter in IsAuthEnabled and IsAuthReady

emlun committed:
Make service.IsAuthEnabled private

emlun committed:
Delete cfg and tokenCookieManager fields from webauthnService

emlun committed:
Move tokenCookieManager and webauthnService back to Serve()

emlun committed:
Revert error result from api.New(...): Service

emlun committed:
Un-capitalize error messages

rasa commented:
@emlun Once this is merged in, I'll retest.

imsodin commented on lib/api/api.go:
```diff
@@ -258,6 +248,15 @@ func (s *service) Serve(ctx context.Context) error {
 	restMux := httprouter.New()
 
+	webauthnDeviceName := ""
+	if dev, ok := s.cfg.Device(s.id); ok {
```
The device name is passed to  the webauthn engine - I assume it's of importance? Because currently if it changes, that change wont propagate. The api service only restarts on gui config changes.

imsodin replied:
Ok, now followup with some actual pointers instead of just pointing out an issue (too tired before): The easiest option would be to just add a check to `api.service.CommitConfiguration` to also restart on device name change. Another option would be to add a mechanism to propagate a device name change from that function to the webauthn service. Or make the webauthnservice "standalone", i.e. let it subscribe to config changes instead of tagging on `api.service` for that (potentially make it even fully independent from api.service, i.e. it's own suture service, to also avoid the restarts - then again restarts when gui config changes might be desirable?).

emlun replied:
Nah, it's not very important at all - it's only used to set the [RP display name](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#dom-publickeycredentialentity-name) when registering credentials. That value may be presented in account picker or credential management UIs, but it's not used for any security decisions.

Actually, it's probably better anyway to just use the device ID rather than the name, because the name can change but the ID is immutable, and as of now there is no way for the RP (Syncthing) to signal to authenticators that the RP's display name should be updated (one might be [coming in WebAuthn L3](https://github.com/w3c/webauthn/pull/2093), but there are authenticators that will never support this anyway). So I think I'll just do that instead and eliminate the problem altogether.

imsodin commented on lib/api/api.go:
```diff
@@ -386,13 +382,14 @@ func (s *service) Serve(ctx context.Context) error {
 	// Wrap everything in auth, if user/password is set or WebAuthn is enabled.
-	if s.IsAuthEnabled() {
-		authMW := newBasicAuthAndSessionMiddleware(&s.tokenCookieManager, guiCfg, s.cfg.LDAP(), handler, s.evLogger)
+	if s.isAuthEnabled(&webauthnService, guiCfg) {
```
Lets just return a pointer to the service from the constructor in the first place. Seems a bit inconsistent to care about copying in some places but not others.

imsodin commented on lib/api/api.go:
```diff
@@ -519,10 +516,9 @@ func (s *service) CommitConfiguration(from, to config.Configuration) bool {
 }
 
-func (s *service) IsAuthEnabled() bool {
+func (s *service) isAuthEnabled(webauthnService *webauthnService, guiCfg config.GUIConfiguration) bool {
```
The receiver is unused now.

emlun replied:
Thanks, fixed now.

imsodin commented on lib/api/api_auth_webauthn.go:
```diff
@@ -126,55 +120,33 @@ func (s *webauthnService) storeState(state config.WebauthnState) error {
-	return s.cfg.GUI().User
+type webauthnLibUser struct {
+	service *webauthnService
+	guiCfg  config.GUIConfiguration
```
The only field you are using is `guiCfg.User`, so you might as well add/pass only that instead of the entire config.

emlun replied:
Not quite - the `webauthnLib.User.WebAuthnCredentials()` method implementation calls `u.service.EligibleWebAuthnCredentials(u.guiCfg)` which references `guiCfg.WebauthnRpId`, so at least those two are both needed.

imsodin replied:
Oh right, thanks for pointing that out.

imsodin reviewed (approved):
Sorry, I started nitting again. Overall looks good. There is one relevant comment about config changes again.

emlun committed:
Add function structutil.WithDefaults

emlun committed:
Set default GUIConfiguration.WebauthnRpId and .WebauthnOrigin in tests

emlun committed:
Remove default rpId in EligibleWebAuthnCredentials

emlun committed:
Remove default reflection from newWebauthnEngine

emlun committed:
Delete unused receiver

emlun committed:
Use short device ID instead of device name in WebAuthn RP display name

This solves 2 problems: no need to reload the name on config changes, because
the device ID should rarely or never change; and consequently we don't get
credentials with a stale RP display name, as there's no way in WebAuthn L2 to
signal to the authenticator that the RP display name (if stored) should be
updated.

emlun commented:
No need to apologize, I really appreciate the comments!

Last night I added a few changes to also address the discussion on the default-reflection in `newWebauthnEngine`:
- https://github.com/syncthing/syncthing/pull/9175#discussion_r1700729175

I considered doing this "globally" in `mocks.Wrapper.GUIReturns()`, and spent an hour or two troubleshooting under the faulty assumption that `structutil.SetDefaults` would set only the properties with a zero value. Turns out it unconditionally overwrites all properties that have a default value tag, which broke pretty much _all_ of the tests in various surprising ways when I applied it that broadly (for example, all those tests competed for port 8384, which de-parallelized most of the test suite so everything timed out). Given that experience, I decided that it's probably better to be explicit about applying these defaults at the construction site than having it happen magically in some hidden shared utility.

@imsodin Would you like to re-review these changes here or shall I just merge them into https://github.com/syncthing/syncthing/pull/9175?

emlun committed:
Clarify behaviour of structutil.SetDefaults in doc comment

imsodin commented:
Lets merge - there are now also a lot of changes here, so the benefit of being small is no longer there and the "change-indirection" adds quite a bit of cognitive load for me :)
